### ISSUE TITLE
Is it possible to use CUDA to accelerate this algorithm?

### ISSUE THREAD

**Issue by wenkaier** [oedokumaci/gale-shapley-algorithm#14]

**Comment by oedokumaci** [comment]
As of now, my implementation of the algorithm does not incorporate CUDA acceleration. My decision was guided by the well-known advice from Donald Knuth: _Premature optimization is the root of all evil_. However, it is entirely feasible to introduce such acceleration by parallelizing the proposal and rejection processes and handling them in batches.

**Comment by wenkaier** [comment]
Thank you very much and endorse your suggestion, I'll try the python version first to see if the performance is up to snuff, and then see if it's necessary to develop a cuda version.

**Comment by oedokumaci** [comment]
Yes, that makes sense. I'd also start with Python. It would be also nice to benchmark performance in tests, I suspect for some number of people _U_, any environment with _2N<U_ people will see a performance **downgrade** because of the overhead. But for large _U_ multiprocessing proposals and responses should see the benefits. I appreciate your effort, thank you very much. I am assigning you to the issue and re-labeling as enhancement.

**Comment by oedokumaci** [comment]
Hey @wenkaier — thanks for the original nudge on this. Circling back with an update.

Just shipped **v1.6.0** with a new optional `numeric` subpackage:

```bash
pip install "gale-shapley-algorithm[numeric]"
```

What you get:

- `numeric.gale_shapley` — fast numpy deferred acceptance on `int16` rank matrices
- `numeric.is_stable` / `is_stable_batch` — single-matching and vectorized stability checks (the batched one filters K candidate matchings in a single tensor op)
- `numeric.enumerate_stable_matchings` — full stable-matching lattice enumeration (up to n=10 via batched brute force)

This covers the "large U" batch story we discussed — for pipelines that run thousands of instances (simulations, ML/RL training, market-design experiments) the vectorized path is dramatically faster than looping `Algorithm.execute()` on `Proposer`/`Responder` objects. Release notes: https://github.com/oedokumaci/gale-shapley-algorithm/releases/tag/v1.6.0.

Going to close this one out for now — the numpy primitives solve the "make it fast at scale" need without pulling CUDA / PyTorch into the project's dependency surface. A real GPU acceleration story (likely a separate `gale_shapley_algorithm.cuda` or `.jax` subpackage) is worth its own issue if you hit limits with the numpy path — happy to chat there if so.

Thanks again for raising this!
